src/dmd/json.d

```diff
@@ -857,17 +871,178 @@ public:
 extern (C++) void json_generate(OutBuffer* buf, Modules* modules)
 {
     scope ToJsonVisitor json = new ToJsonVisitor(buf);
-    json.arrayStart();
-    json.generateCompilerInfo();
-    json.generateBuildInfo();
-    for (size_t i = 0; i < modules.dim; i++)
-    {
-        Module m = (*modules)[i];
-        if (global.params.verbose)
-            message("json gen %s", m.toChars());
-        m.accept(json);
-    }
-    json.generateSemantics();
-    json.arrayEnd();
-    json.removeComma();
+
+    if (global.params.jsonQueryFlags == 0)
+    {
+        // Generate the original "non-query format, which is just an array
+        // of modules representing their syntax.
+        json.generateModules(modules);
+        json.removeComma();
+    }
+    else
+    {
+        // Generate the "query format" which is an object where each
+        // output option is its own field.
+
+        json.objectStart();
+        if (global.params.jsonQueryFlags & JsonQueryFlags.compilerInfo)
+        {
+            json.propertyStart("compilerInfo");
+            json.generateCompilerInfo();
+        }
+        if (global.params.jsonQueryFlags & JsonQueryFlags.buildInfo)
+        {
+            json.propertyStart("buildInfo");
+            json.generateBuildInfo();
+        }
+        if (global.params.jsonQueryFlags & JsonQueryFlags.modules)
+        {
+            json.propertyStart("modules");
+            json.generateModules(modules);
+        }
+        if (global.params.jsonQueryFlags & JsonQueryFlags.semantics)
+        {
+            json.propertyStart("semantics");
+            json.generateSemantics();
+        }
+        json.objectEnd();
+    }
+}
+
+/**
+Each flag represents a field that can be included in the JSON output.
+*/
+private enum JsonQueryFlags
+{
+    compilerInfo = 0x01,
+    buildInfo    = 0x02,
+    modules      = 0x04,
+    semantics    = 0x08,
+}
+private enum jsonFieldNames = () {
+    string s;
+    string prefix = "";
+    foreach (enumName; __traits(allMembers, JsonQueryFlags))
+    {
+        s ~= prefix ~ enumName;
+        prefix = ", ";
+    }
+    s ~= "\0"; // make sure it is null-terminated
+    return s;
+}();
+
+/**
+Parses a JSON query into a set of flags.
+
+Returns: false on error, prints its own error messages.
+*/
+bool parseJsonQuery(const(char)* query, uint* outQueryFlags)
+{
+    auto parser = QueryParser(query);
```

Comment:
This seems like a lot of effort to simply turn a comma-separated list of 4 identifiers into 4 flags? In fact, this could be avoided entirely by just supporting multiple -Xq=identifier queries, which you already do:
```
dmd -Xq=compilerInfo,buildInfo,semantics,modules
```
becomes:
```
dmd -Xq=compilerInfo -Xq=buildInfo -Xq=semantics -Xq=modules
```
and all this code goes away. Note we do the same thing for multiple `-version=identifier` switches, etc.

Reply:
there's a precedent for that with:
* `-i=+core,-std.stdio`   
* and in clang, `-Wl,-lfoo,-lbar`
* and the suggested -unittest=std.stdio,foo.bar

maybe exposing that logic (parsing comma separated list) into reusable code would be an option

Reply:
I would have argued against the -i syntax, on the same grounds. What clang does is not that compelling.

In any case, reinventing another parser for CSV options is not the best plan. As noted in the comments, dmd already has several different ways of specifying a file, which is a miserably stupid design, and is really hard to unwind because of subtle breakage. Let's not do that with multiple CSV parsers with same-but-maddeningly-different behaviors.

Reply:
If you prefer using multiple arguments, i.e.
```
-i=foo -i=-foo.bar -i=std -i=core
```
 I don't think it's too late to change `-i`.  It hasn't showed up in a major release yet.  I agree that consistency is virtue we should aspire to.  Just say the word and I'll create a PR for this.

Reply:
I've created a PR to remove the "comma support" for `-i`: https://github.com/dlang/dmd/pull/7857/files